src/auctions.lua

```diff
@@ -0,0 +1,91 @@
+local Auction = {}
+
+-- Default Auction Settings
+AuctionSettings = {
+	durationMs = 60 * 1000 * 60 * 24 * 14, -- 14 days in milliseconds
+	decayRate = 0.000000000016847809193121693, -- 0.02037911 / durationMs
+	scalingExponent = 190, -- steepness of the curve
+	startPriceMultiplier = 50, -- multiplier for the starting price
+}
+
+--- Represents an Auction.
+--- @class Auction
+--- @field name string The name of the auction
+--- @field demandFactor number The demand factor for pricing
+--- @field baseFee number The base fee for the auction
+--- @field initiator string The address of the initiator of the auction
+--- @field settings table The settings for the auction
+--- @field startTimestamp number The starting timestamp for the auction
+--- @field endTimestamp number The ending timestamp for the auction
+--- @field registrationFeeCalculator function Function to calculate registration fee
+
+--- Creates a new Auction instance
+--- @param name string The name of the auction
+--- @param startTimestamp number The starting timestamp for the auction
+--- @param demandFactor number The demand factor for pricing
+--- @param baseFee number The base fee for the auction
+--- @param initiator string The address of the initiator of the auction
+--- @param registrationFeeCalculator function Function to calculate registration fee that supports type, baseFee, years, demandFactor
+--- @return Auction The new Auction instance
+function Auction:new(name, startTimestamp, demandFactor, baseFee, initiator, registrationFeeCalculator)
+	local auction = {
+		name = name,
+		initiator = initiator,
+		startTimestamp = startTimestamp,
+		endTimestamp = startTimestamp + AuctionSettings.durationMs,
+		registrationFeeCalculator = registrationFeeCalculator,
+		baseFee = baseFee,
+		demandFactor = demandFactor,
+		settings = {
+			durationMs = AuctionSettings.durationMs,
+			decayRate = AuctionSettings.decayRate,
+			scalingExponent = AuctionSettings.scalingExponent,
+			startPriceMultiplier = AuctionSettings.startPriceMultiplier,
+		},
+	}
+	setmetatable(auction, self)
+	self.__index = self
```

Comment:
Seems there's some nuances to setting up the class this way (in the constructor) vs. separately. Will share chatGPT thread with you separately about it.